lib/Pipes/PipelineManager.cpp

```diff
@@ -162,11 +155,124 @@ checkComponentsVersion(const revng::DirectoryPath &ExecutionDirectory,
+
+  if (auto Error = ModelFile.copyTo(ModelBackupFile.get()); !!Error)
+    return Error;
+
+  return ModelBackupFile;
+}
+
+static llvm::Error migrateModelFile(const revng::FilePath &ModelFile) {
+  TemporaryFile InputTemporaryFile("revng-model-migration-input", "yml");
+  TemporaryFile OutputTemporaryFile("revng-model-migration-output", "yml");
+
+  revng::FilePath
+    InputFile = revng::FilePath::fromLocalStorage(InputTemporaryFile.path());
+  if (auto Error = ModelFile.copyTo(InputFile); !!Error)
+    return Error;
+
+  {
+    int ReturnCode = ::Runner.run("revng",
+                                  { "model",
+                                    "migrate",
+                                    "--input",
+                                    InputTemporaryFile.path().str(),
+                                    "--output",
+                                    OutputTemporaryFile.path().str() });
+
+    if (ReturnCode != 0)
+      return revng::createError("Error encountered while running migrations, "
+                                "process returned: %d",
+                                ReturnCode);
+  }
+
+  revng::FilePath
+    OutputFile = revng::FilePath::fromLocalStorage(OutputTemporaryFile.path());
+  if (auto Error = OutputFile.copyTo(ModelFile); !!Error)
+    return Error;
+
+  return llvm::Error::success();
+}
+
+// Conveninece function for setUpPipeline, creates a backup for the model file
+// and runs migrations on it. On success returns the backup path, on failure
+// returns the error and restores the model file from backup. If something goes
+// horribly wrong and it fails to restore, it aborts.
+static llvm::Expected<revng::FilePath>
+migrateModel(const revng::DirectoryPath &ExecutionDirectory) {
+  auto ModelFile = ExecutionDirectory.getDirectory("context")
+                     .getFile(revng::ModelGlobalName);
+
+  if (not ModelFile.exists()) {
+    // Model file does not exist, so no migrations will actually run (empty
+    // model is assumed to be in the latest version). So for the sake of better
+    // error reporting, process this case explicitly.
+    return revng::createError("Migrations will not be run since the model file "
+                              "does not exist");
+  }
+
+  llvm::Expected<revng::FilePath>
+    ModelBackupFile = createModelBackup(ModelFile);
+
+  if (not ModelBackupFile)
+    return ModelBackupFile.takeError();
+
+  if (auto MigrationError = migrateModelFile(ModelFile); !!MigrationError) {
+    // Migration failed, rollback to the previous version of the model.
+    if (auto CopyError = ModelBackupFile->copyTo(ModelFile); !!CopyError) {
+      // BIG PROBLEM: model likely changed and retrieval from backup failed.
+      revng_abort("Model migration failed and the model could not be rolled "
+                  "back to the original version, this requires some manual "
+                  "intervention. The model backups can be found alongside the "
+                  "model file.");
```

Comment:
Hmmm, so we end up here if the following happens:
1. Model parsing fails
2. We back up the model
3. We try to upgrade the model via `revng model migrate`, which fails
4. We try to restore the model from (2) but it fails

I think there should be some marker file like `context/.broken-model` or something similar that prevents successive startups from proceeding.
What do you think @aleclearmind ?

Reply:
Now that I think about it, we should probably do something like:

1. Copy model file to another file
2. Migrate the other file, not the original model
3. If it fails, return with failure
4. If it succeeds, swap the two files

You could technically still end up with a broken model file if swapping fails, but that's much less likely.

Reply:
The model gets copied anyway because it needs to be local. So without actually backing up stuff I'd do:
1. Copy model to the temporary file
2. Try and run migrations, if they fail return right away
3. Copy the temporary output file back to the model file

Reply:
That's what I meant

Reply:
Now it works as you mentioned. There was no change to the logic, I only chopped off the restore-from-backup-if-failed step without additional risk. Let me know if it looks okay now.

Reply:
Yes, that's ok